DotaStats/Modules/SearchPlayerModule/SearchPlayerModulePresenter.swift

```diff
@@ -28,64 +35,93 @@ final class SearchPlayerModulePresenter {
             return .startScreen
         case .loading:
             return .loading
-        case .result(let result):
-            switch result {
-            case .success:
-                return players.isEmpty ? .empty : .success
-            case .failure:
-                return .failure
-            }
+        case .success(let players):
+            return players.isEmpty ? .empty : .success
+        case .failure:
+            return .failure
         }
     }
     
     private var state: SearchPlayerModulePresenterState {
         didSet {
             oldValue.token?.cancel()
             switch state {
-            case .result(let requestResult):
-                switch requestResult {
-                case .success(let players):
-                    self.players = players
-                case .failure(_):
-                    break
-                }
-            case .none, .loading:
+            case .success(let result):
+                players = result
+            case .none, .loading, .failure:
                 break
             }
             
             view?.updateState(viewState)
         }
     }
     
-    init(output: SearchPlayerModuleOutput, networkService: PlayerSearchNetworkService) {
+    init(output: SearchPlayerModuleOutput, playerSearchService: PlayerSearchService, imageNetworkService: ImageNetworkService) {
         self.output = output
-        self.networkService = networkService
+        self.playerSearchService = playerSearchService
+        self.imageNetworkService = imageNetworkService
         state = .none
     }
+    
+    private func cancelAllLoading() {
+        imageRequestTokens.forEach { _, request in
+            request.cancel()
+        }
+    }
+    
+    deinit {
+        cancelAllLoading()
+    }
 }
 
 // MARK: - SearchModuleViewOutput
 
 extension SearchPlayerModulePresenter: SearchPlayerModuleViewOutput {
-    var count: Int {
+    var countOfRows: Int {
         players.count
     }
     
-    func getData(indexPath: IndexPath) -> Players {
-        players[indexPath.row]
+    func getData(indexPath: IndexPath) -> PlayerInfoFromSearch {
+        if let urlString = players[indexPath.row].avatarFull,
```

Comment:
let player = players[indexPath.row]

loadPlayerAvatar(player: ***)  { [weak self] in
 self?.view.reload(at: indexPath)
}

return player


func loadPlayerAvatar(player: Player., completion: @escaping () -> Void) {
    guard player.avatar == nil,
             let urlString = player.avatarFull,
             let url = URL(string: urlString) else {
                return 
     }

    let imageRequestToken = imageNetworkService.load ...
      player.avatar = image
   }

}